### ISSUE TITLE
Cannot parse Chrome Bookmarks file, and _iterate_folder_html doesn't appear to be recursive

### ISSUE THREAD

**Issue by phx** [radam9/bookmarks-converter#112]
`core.py` looks for a list in the `roots` dict if the browser is Chrome.  However, Bookmarks folder lookes like this:

```
{ 
  "checksum": "example",
  "roots": {
    "bookmark_bar": { "name": "Bookmarks bar",  "id": "1", "children": [(stuff)], ...},
    "other": { "name": "Other Bookmarks", "id": "2", "children": [], ...},
    "synced": { "name": "Mobile Bookmarks", "id": "3", "children": [], ...}
  }
}
```

Trying to parse this file yields an error, because like it said, it's looking for a list instead of a dict.

However, if I try to pass in `roots['bookmark_bar']['children']` as the actual JSON to parse, this "works", but the html output is nothing that can be imported in any logical order because of a possible oversight (?) in `core.py` in the `_iterate_folder_html` method.  A short snippet of the final html output I get is as follows (note that my "categories" folder in my "Bookmarks bar" is quite extensive with lots of nested folders):

```
<TITLE>Bookmarks</TITLE>
--
  | <H1>Bookmarks Menu</H1>
  |  
  | <DL><p>
  | <DT><H3 ADD_DATE="1655413626" LAST_MODIFIED="0" PERSONAL_TOOLBAR_FOLDER="true">Bookmarks bar</H3>
  | <DL><p>
  | <DT><H3 ADD_DATE="1655413626" LAST_MODIFIED="0">categories</H3>
  | <DL><p>
  | <folder4><folder5><folder6><folder7><folder10><folder14><folder15><folder16><folder17><folder18><folder19><folder20><folder22><folder23><folder76><folder94><folder95><folder122><folder123><folder124><folder125><folder126><folder128><folder132><folder133><folder135><folder136><folder137><folder138><folder139><folder140><DT><H3 ADD_DATE="1655413626" LAST_MODIFIED="0">mon</H3>
  | <DL><p>
  | <DT><H3 ADD_DATE="1655413626" LAST_MODIFIED="0">alertsite</H3>
  | <DL><p>
  | <DT><A HREF="http://www.alertsite.com/cgi-bin/helpme.cgi?page=monitoring_locations.html" ADD_DATE="1655413626" LAST_MODIFIED="0" ICON_URI="None" ICON="None">AlertSite Monitoring Locations-IPs</A>
  | </DL><p>
```

Based on the code, it is my understanding that these placeholders are supposed to get replaced with the actual information that resides in the bookmark stack (which is absolutely correct when viewed with a debugger), but they don't seem to be getting replaced or are overlooked somehow from non-recursion or something.

That being said, initially parsing a `bookmarks.html` file into the Bookmarkie formatted JSON, and subsequently parsing back to HTML works just fine.

The trouble seems to happen when parsing the native Chrome Bookmarks JSON file, where things seem to get all out of order.

**Comment by phx** [comment]
I am leaving a few somewhat-sanitized files in a zip file to try and help replicate and troubleshoot this issue:

- `ChromeBookmarks.example.json` - an example of my raw Chrome Bookmarks file
- `bookmarks.html` - an example of what the result should look like (exported directly from Chrome)
- `bookmarks.convertedfromhtml.json` - the resulting (good) JSON file from converting from `bookmarks.html` and can be re-converted to HTML without losing data.

(The problem comes from trying to parse the `ChromeBookmarks.example.json`, which is a direct representation of the Chrome Bookmarks file.)

[bookmarks_troubleshooting.zip](https://github.com/radam9/bookmarks-converter/files/8929920/bookmarks_troubleshooting.zip)

**Comment by radam9** [comment]
In the zip file you have attached you have provided a `ChromeBookmarks.example.json` as a source file.
I have checked the file and that is not a standard chrome json file.
First of all `bookmark_bar`, `other` and `synced` are in the root of the dictionary in the file, they should have been in a `"roots": {}` element.

you have provided:
```json
{ 
  "checksum": "example",
  "bookmark_bar": {   <--- note there is no "roots" element here
    "children": ["stuff inside here"],
    "date_added": "13284592559028856",
    "date_modified": "13299254725932084",
    "guid": "0bc5d13f-2cba-5d74-951f-3f233fe6c908",
    "id": "1",
    "name": "Bookmarks bar",
    "type": "folder"
  },
  "other": {
    "children": [],
    "date_added": "13284592559028858",
    "date_modified": "13269200466599596",
    "guid": "82b081ec-3dd3-529c-8475-ab6c344590dd",
    "id": "2",
    "name": "Other bookmarks",
    "type": "folder"
  },
  "synced": {
    "children": [],
    "date_added": "13284592559028861",
    "date_modified": "0",
    "guid": "4cf2e351-0e85-532b-bb37-df045d8f8d0f",
    "id": "3",
    "name": "Mobile bookmarks",
    "type": "folder"
  }
}
```
it should be
```json
{
    "checksum": "example",
    "roots": {     <-- notice that `bookmark_bar` and the other items should be inside "roots"
        "bookmark_bar": {
            "children": [
                "stuff inside here"
            ],
            "date_added": "13284592559028856",
            "date_modified": "13299254725932084",
            "guid": "0bc5d13f-2cba-5d74-951f-3f233fe6c908",
            "id": "1",
            "name": "Bookmarks bar",
            "type": "folder"
        },
        "other": {
            "children": [],
            "date_added": "13284592559028858",
            "date_modified": "13269200466599596",
            "guid": "82b081ec-3dd3-529c-8475-ab6c344590dd",
            "id": "2",
            "name": "Other bookmarks",
            "type": "folder"
        },
        "synced": {
            "children": [],
            "date_added": "13284592559028861",
            "date_modified": "0",
            "guid": "4cf2e351-0e85-532b-bb37-df045d8f8d0f",
            "id": "3",
            "name": "Mobile bookmarks",
            "type": "folder"
        }
    }
}
```

then I also found items that look like this:
```json
{
    "date_added": "13192729572715308",
    "guid": "0a412f82-1ac2-4fe1-8c1b-7301ba6a0805",
    "id": "93",
    "meta_info": {
        "last_visited_desktop": "13204918293394216"
    },
    "name": "VMware ESXi",
    "type": "url",
    "url": "https://www.google.com"
}
```
`meta_info` is not a field that exists in standard chrome json files.
The nested folders are not an issue as I have parsed the file shown below without an issue.
```json
{
   "checksum": "8b092e15e4830edc6bcfbbc48d8a6a61",
   "roots": {
      "bookmark_bar": {
         "children": [  ],
         "date_added": "13299965534893698",
         "date_modified": "0",
         "guid": "0bc5d13f-2cba-5d74-951f-3f233fe6c908",
         "id": "1",
         "name": "Bookmarks bar",
         "type": "folder"
      },
      "other": {
         "children": [ {
            "children": [ {
               "children": [ {
                  "children": [ {
                     "children": [ {
                        "children": [ {
                           "date_added": "13299965596150595",
                           "guid": "bf09801a-0e04-4e0a-a0d8-1faadf33d34f",
                           "id": "10",
                           "name": "Google",
                           "type": "url",
                           "url": "https://www.google.com/"
                        } ],
                        "date_added": "13299965585971078",
                        "date_modified": "13299965599677604",
                        "guid": "28983b6f-f135-4560-aa41-12fb755ac445",
                        "id": "9",
                        "name": "folder5",
                        "type": "folder"
                     } ],
                     "date_added": "13299965577171476",
                     "date_modified": "13299965596150595",
                     "guid": "2a273398-6f69-40f9-8d2c-393b770c86ba",
                     "id": "8",
                     "name": "folder4",
                     "type": "folder"
                  } ],
                  "date_added": "13299965570447717",
                  "date_modified": "13299965577171541",
                  "guid": "be4212bc-158d-47b5-b9a1-4b1dbd0b5114",
                  "id": "7",
                  "name": "folder3",
                  "type": "folder"
               } ],
               "date_added": "13299965565263549",
               "date_modified": "13299965570447856",
               "guid": "f68bd3bf-9336-4ee7-a475-e169005586e5",
               "id": "6",
               "name": "folder2",
               "type": "folder"
            } ],
            "date_added": "13299965561080757",
            "date_modified": "13299965565263612",
            "guid": "74b9f0a6-c6b4-42ee-9310-4dc2222078d8",
            "id": "5",
            "name": "folder1",
            "type": "folder"
         }, {
            "children": [ {
               "children": [ {
                  "children": [ {
                     "children": [ {
                        "children": [  ],
                        "date_added": "13299965661436453",
                        "date_modified": "13299965661436453",
                        "guid": "236da8d8-1c25-4edc-b036-7cff5784131f",
                        "id": "16",
                        "name": "folder_e",
                        "type": "folder"
                     } ],
                     "date_added": "13299965657280981",
                     "date_modified": "13299965661436584",
                     "guid": "7da85534-03e4-4aa2-8b1f-b4e1c9aaac9d",
                     "id": "15",
                     "name": "folder_d",
                     "type": "folder"
                  } ],
                  "date_added": "13299965653556319",
                  "date_modified": "13299965657281040",
                  "guid": "839fa367-13a5-4e2b-a71f-19c80e77e133",
                  "id": "14",
                  "name": "folder_c",
                  "type": "folder"
               } ],
               "date_added": "13299965647147112",
               "date_modified": "13299965653556407",
               "guid": "c8c7c5ea-24f8-4bc9-9706-3dcb648c0271",
               "id": "13",
               "name": "folder_b",
               "type": "folder"
            } ],
            "date_added": "13299965642644876",
            "date_modified": "13299965647147183",
            "guid": "e7b53982-86e9-417b-9289-65fc2ca3b3dc",
            "id": "12",
            "name": "folder_a",
            "type": "folder"
         } ],
         "date_added": "13299965534893700",
         "date_modified": "13299965642645007",
         "guid": "82b081ec-3dd3-529c-8475-ab6c344590dd",
         "id": "2",
         "name": "Other bookmarks",
         "type": "folder"
      },
      "synced": {
         "children": [  ],
         "date_added": "13299965534893700",
         "date_modified": "0",
         "guid": "4cf2e351-0e85-532b-bb37-df045d8f8d0f",
         "id": "3",
         "name": "Mobile bookmarks",
         "type": "folder"
      }
   },
   "version": 1
}

```

**Comment by phx** [comment]
Sorry I have included the roots in this new upload. I'm trying to sanitize the URLs, which comes with a bit of difficulty.
[bookmarks_troubleshooting.zip](https://github.com/radam9/bookmarks-converter/files/8930956/bookmarks_troubleshooting.zip)

**Comment by phx** [comment]
As you can see, even when I copy the raw file, it does not appear to work from cli:

```
![image](https://user-images.githubusercontent.com/4684896/174391042-3ef0f5c6-48bc-42d4-b3e5-cfa3dcb7a019.png)
```

**Comment by radam9** [comment]
yes, that is because the `ChromeBookmarks.example.json` file you provided in the second zip file contains the `meta_info` fields which are not part of the chrome json standard
```json
{
    "date_added": "13192729572715308",
    "guid": "0a412f82-1ac2-4fe1-8c1b-7301ba6a0805",
    "id": "93",
    "meta_info": {
        "last_visited_desktop": "13204918293394216"
    },
    "name": "VMware ESXi",
    "type": "url",
    "url": "https://www.google.com"
}
```

**Comment by phx** [comment]
How could this info have gotten into my Bookmarks file, and how would you suggest removing it?

**Comment by radam9** [comment]
well, if chrome has not added this new field recently (which is something I will need to look into when I get some time), I would guess that the json you are trying to parse is not the standard chrome json, but rather a json file that has been modified or exported by some other software.
as a quick solution you can manually remove the `meta_info` field from the file (I see that there are 31 entries).

**Comment by phx** [comment]
I have removed the `meta_info` from the file, but html exports still get imported in a wonky fashion.

**Comment by phx** [comment]
I removed the `meta_info` manually from the `Bookmarks` file, and it worked like a charm.  I have seen instances of this in other Bookmarks files, so it may have become part of the spec.  Is there any way that you could update the code to automatically remove the `meta_info` before proceeding?  That's all that I need, and it will work perfectly.

**Comment by phx** [comment]
I have been unable to do this programmatically using Python thus far, as I can't seem to figure out the recursion. But when I did it manually, it worked great.  It would be awesome if your package could account for this automatically so that everything just works flawlessly.

**Comment by radam9** [comment]
A fix has been released in version 0.3.2, that should import the chrome json you provided without any issues.

**Comment by phx** [comment]
Tested. Works perfectly.
